### PR TITLE
[SYCLomatic] Fix the migration of cudaPointerAttributes.type/memoryType by refactoring the MemberExprRewriter

### PR DESCRIPTION
**Pull request by zhiweij1**
Signed-off-by: Jiang, Zhiwei <zhiwei.jiang@intel.com>

### CODE REVIEW

**Review by ShengchenJ — APPROVED** [review]

File: clang/lib/DPCT/MemberExprRewriter.cpp [hunk 1]
```diff
@@ -7,158 +7,350 @@
...
+public:
+  MELiteralNumberMigration(
+      std::shared_ptr<MemberExprRewriterFactoryBase> InnerFactory)
+      : Inner(InnerFactory) {}
+  std::shared_ptr<MERewriter> create(const MemberExpr *ME) const override {
+    if (const IntegerLiteral *IL = isComparingWithIntegerLiteral(ME)) {
+      Expr::EvalResult Result{};
+      IL->EvaluateAsInt(Result, DpctGlobalInfo::getContext());
+      int64_t Value = Result.Val.getInt().getExtValue();
+      std::string Repl;
```
Comment (zhiweij1) [inline clang/lib/DPCT/MemberExprRewriter.cpp:149]:
```suggestion
      std::string Repl = ExprAnalysis::ref(IL);
```

File: clang/lib/DPCT/MemberExprRewriter.cpp [hunk 2]
```diff
@@ -7,158 +7,350 @@
...
+        Repl = MapNames::getClNamespace() + "usm::alloc::host";
+        break;
+      case 2:
+        Repl = MapNames::getClNamespace() + "usm::alloc::device";
+        break;
+      case 3:
+        Repl = MapNames::getClNamespace() + "usm::alloc::shared";
+        break;
+      }
+      DpctGlobalInfo::getInstance().addReplacement(
```
Comment (zhiweij1) [inline clang/lib/DPCT/MemberExprRewriter.cpp]:
```suggestion
      if (!Repl.empty())
        DpctGlobalInfo::getInstance().addReplacement(
```

File: clang/lib/DPCT/MemberExprRewriter.cpp [hunk 3]
```diff
@@ -7,158 +7,351 @@
...
-    if (Base->getType()->isPointerType())
-      return true;
-    return false;
-  };
-}
+  const IntegerLiteral *
+  isComparingWithIntegerLiteral(const MemberExpr *ME) const {
+    const Expr *Cur = ME;
+    const Expr *Pre = nullptr;
+    while (const Expr *E = DpctGlobalInfo::findParent<Expr>(Cur)) {
```
Comment (zhiweij1) [inline clang/lib/DPCT/MemberExprRewriter.cpp]: pass a lambda to find parent

File: clang/lib/DPCT/MemberExprRewriter.cpp [hunk 4]
```diff
@@ -7,158 +7,351 @@
...
+  }
 
-std::function<std::string(const MemberExpr *)>
-                makeMemberGetCall(std::string FuncName) {
-  return [=] (const MemberExpr * ME) -> std::string {
-    return FuncName + "()";
-  };
-}
+public:
+  MELiteralNumberMigration(
```
Comment (zhiweij1) [inline clang/lib/DPCT/MemberExprRewriter.cpp]: use a special name or pass the map into it

**Review by ziranzha — APPROVED** [review]

**Pull request merged**